File: main/java/com/example/wofertas/MinhasOfertas.java
// app/src/main/java/com/example/wofertas/MinhasOfertas.java
package com.example.wofertas;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.util.Log;
import android.view.MenuItem;
import android.view.View;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.bottomnavigation.BottomNavigationView;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.Query;
import com.google.firebase.storage.FirebaseStorage;
import com.google.firebase.storage.StorageReference;

import java.util.ArrayList;
import java.util.List;

public class MinhasOfertas extends AppCompatActivity
        implements SupermercadoOfertaAdapter.OnItemActionListener { // <--- Implementa a interface correta

    private static final String TAG = "MinhasOfertas";

    private RecyclerView recyclerView;
    private SupermercadoOfertaAdapter adapter; // <--- Usando o SupermercadoOfertaAdapter
    private List<Oferta> listaOfertas;
    private ProgressBar progressBar;
    private TextView tvNoOffersMessage; // Mensagem de "nenhuma oferta"
    private BottomNavigationView bottomNavigationView;
    private Toolbar toolbar;

    private FirebaseAuth mAuth;
    private FirebaseUser currentUser;
    private FirebaseFirestore db;
    private FirebaseStorage storage;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_minhas_ofertas); // Certifique-se de ter este layout

        toolbar = findViewById(R.id.toolbar_minhas_ofertas); // ID do toolbar
        setSupportActionBar(toolbar);
        if (getSupportActionBar() != null) {
            getSupportActionBar().setTitle("Minhas Ofertas");
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
            toolbar.setNavigationOnClickListener(v -> finish());
        }

        recyclerView = findViewById(R.id.recyclerViewMinhasOfertas); // ID do RecyclerView
        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        listaOfertas = new ArrayList<>();
        // Instanciando o SupermercadoOfertaAdapter e passando 'this' como listener
        adapter = new SupermercadoOfertaAdapter(this, listaOfertas, this);
        recyclerView.setAdapter(adapter);

        progressBar = findViewById(R.id.progressBarMinhasOfertas); // ID do ProgressBar
        tvNoOffersMessage = findViewById(R.id.tvNoOffersMessage); // ID da TextView de mensagem
        bottomNavigationView = findViewById(R.id.bottom_navigation_supermercado); // ID da BottomNavigationView

        mAuth = FirebaseAuth.getInstance();
        db = FirebaseFirestore.getInstance();
        storage = FirebaseStorage.getInstance();
        currentUser = mAuth.getCurrentUser();

        if (currentUser == null) {
            Toast.makeText(this, "Usuário não autenticado. Redirecionando para o login.", Toast.LENGTH_LONG).show();
            Intent intent = new Intent(this, LoginActivity.class);
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
            startActivity(intent);
            finish();
            return;
        }

        setupBottomNavigationView();
    }

    @Override
    protected void onResume() {
        super.onResume();
        carregarMinhasOfertas();
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            onBackPressed();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    /**
     * Configura a navegação inferior (BottomNavigationView) para o perfil do supermercado.
     */
    private void setupBottomNavigationView() {
        bottomNavigationView.setSelectedItemId(R.id.navigation_gerenciar_ofertas); // Define este item como selecionado
        bottomNavigationView.setOnItemSelectedListener(item -> {
            int id = item.getItemId();
            if (id == R.id.navigation_publicar_oferta) {
                startActivity(new Intent(this, DashboardSupermercadoActivity.class));
                finish();
                return true;
            } else if (id == R.id.navigation_dashboard_supermercado) {
                // Se você tiver uma tela de "dashboard" mais geral, pode ir para ela
                startActivity(new Intent(this, GerenciarOfertasSupermercadoActivity.class));
                finish();
                return true;
            } else if (id == R.id.navigation_gerenciar_ofertas) {
                Toast.makeText(this, "Você já está em Minhas Ofertas.", Toast.LENGTH_SHORT).show();
                return true; // Já está nesta Activity
            } else if (id == R.id.navigation_perfil_supermercado) {
                startActivity(new Intent(this, Perfil.class));
                finish();
                return true;
            }
            return false;
        });
    }

    /**
     * Carrega as ofertas publicadas pelo supermercado logado.
     */
    private void carregarMinhasOfertas() {
        if (currentUser == null) {
            return;
        }

        progressBar.setVisibility(View.VISIBLE);
        tvNoOffersMessage.setVisibility(View.GONE); // Esconde a mensagem enquanto carrega

        db.collection("ofertas")
                .whereEqualTo("supermercadoId", currentUser.getUid())
                .orderBy("timestamp", Query.Direction.DESCENDING)
                .get()
                .addOnCompleteListener(task -> {
                    progressBar.setVisibility(View.GONE);
                    if (task.isSuccessful()) {
                        listaOfertas.clear();
                        for (com.google.firebase.firestore.QueryDocumentSnapshot document : task.getResult()) {
                            Oferta oferta = document.toObject(Oferta.class);
                            oferta.setOfertaId(document.getId());
                            listaOfertas.add(oferta);
                        }
                        adapter.atualizarLista(listaOfertas);
                        if (listaOfertas.isEmpty()) {
                            tvNoOffersMessage.setVisibility(View.VISIBLE);
                        } else {
                            tvNoOffersMessage.setVisibility(View.GONE);
                        }
                    } else {
                        Log.w(TAG, "Erro ao carregar minhas ofertas.", task.getException());
                        Toast.makeText(this, "Erro ao carregar suas ofertas.", Toast.LENGTH_SHORT).show();
                        tvNoOffersMessage.setVisibility(View.VISIBLE);
                    }
                });
    }

    /**
     * Implementação da interface SupermercadoOfertaAdapter.OnItemActionListener
     * Lida com o clique no botão "Ver PDF".
     */
    @Override
    public void onVerPdfClick(Oferta oferta) {
        if (oferta.getUrlPdf() != null && !oferta.getUrlPdf().isEmpty()) {
            Intent browserIntent = new Intent(Intent.ACTION_VIEW, Uri.parse(oferta.getUrlPdf()));
            startActivity(browserIntent);
        } else {
            Toast.makeText(this, "PDF da oferta não disponível.", Toast.LENGTH_SHORT).show();
        }
    }


    /**
     * Implementação da interface SupermercadoOfertaAdapter.OnItemActionListener
     * Lida com o clique no botão de editar oferta.
     */
    @Override
    public void onEditClick(Oferta oferta) {
        Toast.makeText(this, "Editando oferta: " + oferta.getDescricao(), Toast.LENGTH_SHORT).show();
        Intent intent = new Intent(this, DashboardSupermercadoActivity.class);
        intent.putExtra("ofertaParaEditar", oferta);
        startActivity(intent);
    }


    /**
     * Implementação da interface SupermercadoOfertaAdapter.OnItemActionListener
     * Lida com o clique no botão de excluir oferta.
     */
    @Override
    public void onDeleteClick(Oferta oferta, int position) {
        new AlertDialog.Builder(this)
                .setTitle("Excluir Oferta")
                .setMessage("Tem certeza que deseja excluir a oferta '" + oferta.getDescricao() + "'?")
                .setPositiveButton("Excluir", (dialog, which) -> excluirOferta(oferta, position))
                .setNegativeButton("Cancelar", null)
                .show();
    }

    /**
     * Exclui a oferta do Firestore e o PDF associado do Storage.
     * @param oferta A oferta a ser excluída.
     * @param position A posição da oferta na lista do adaptador.
     */
    private void excluirOferta(Oferta oferta, int position) {
        progressBar.setVisibility(View.VISIBLE);

        // 1. Excluir o PDF do Firebase Storage (se houver)
        if (oferta.getUrlPdf() != null && !oferta.getUrlPdf().isEmpty()) {
            try {
                StorageReference pdfRef = storage.getReferenceFromUrl(oferta.getUrlPdf());
                pdfRef.delete()
                        .addOnSuccessListener(aVoid -> Log.d(TAG, "PDF excluído do Storage: " + oferta.getUrlPdf()))
                        .addOnFailureListener(e -> Log.w(TAG, "Falha ao excluir PDF do Storage: " + e.getMessage()));
            } catch (IllegalArgumentException e) {
                Log.e(TAG, "URL do PDF inválida para exclusão: " + oferta.getUrlPdf() + " - " + e.getMessage());
            }
        }

        // 2. Excluir o documento da oferta do Firestore
        db.collection("ofertas").document(oferta.getOfertaId())
                .delete()
                .addOnSuccessListener(aVoid -> {
                    Toast.makeText(this, "Oferta excluída com sucesso!", Toast.LENGTH_SHORT).show();
                    listaOfertas.remove(position);
                    adapter.notifyItemRemoved(position);
                    progressBar.setVisibility(View.GONE);
                    if (listaOfertas.isEmpty()) {
                        tvNoOffersMessage.setVisibility(View.VISIBLE);
                    }
                })
                .addOnFailureListener(e -> {
                    Log.w(TAG, "Erro ao excluir oferta do Firestore.", e);
                    Toast.makeText(this, "Erro ao excluir oferta: " + e.getMessage(), Toast.LENGTH_SHORT).show();
                    progressBar.setVisibility(View.GONE);
                });
    }
}

File: app/src/main/java/com/example/wofertas/MinhasOfertas.kt
package com.example.wofertas

import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.MenuItem
import android.view.View
import android.widget.ProgressBar
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.widget.Toolbar
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.wofertas.network.ApiClient
import com.example.wofertas.ui.publicar.PublicarOfertaActivity
import com.example.wofertas.ui.encartes.EncartesActivity
import com.google.android.material.bottomnavigation.BottomNavigationView
import kotlinx.coroutines.launch

/**
 * Gerenciamento de ofertas pelo Supermercado.
 * Permite visualizar, editar e excluir ofertas próprias.
 */
class MinhasOfertas : BaseMercadoActivity(), SupermercadoOfertaAdapter.OnItemActionListener {

    private lateinit var recyclerView: RecyclerView
    private lateinit var ofertaAdapter: SupermercadoOfertaAdapter
    private lateinit var progressBar: ProgressBar
    private lateinit var tvMessage: TextView
    private lateinit var bottomNavigationView: BottomNavigationView

    private val listaOfertas: MutableList<Oferta> = mutableListOf()
    private val api get() = ApiClient.authService(this)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_minhas_ofertas)

        val toolbar = findViewById<Toolbar>(R.id.toolbar_minhas_ofertas)
        setSupportActionBar(toolbar)
        supportActionBar?.apply {
            title = "Minhas Ofertas"
            setDisplayHomeAsUpEnabled(true)
        }
        toolbar.setNavigationOnClickListener { finish() }

        recyclerView          = findViewById(R.id.recyclerViewMinhasOfertas)
        progressBar           = findViewById(R.id.progressBarMinhasOfertas)
        tvMessage             = findViewById(R.id.tvMinhasOfertasMessage)
        bottomNavigationView  = findViewById(R.id.bottom_navigation_supermercado)

        // IMPORTANTE: Usando SupermercadoOfertaAdapter para ter botões de Edição/Exclusão
        ofertaAdapter = SupermercadoOfertaAdapter(mutableListOf(), this, this)
        recyclerView.layoutManager = LinearLayoutManager(this)
        recyclerView.adapter = ofertaAdapter

        configurarBottomNav()
        carregarOfertas()
    }

    override fun onResume() {
        super.onResume()
        carregarOfertas()
    }

    private fun carregarOfertas() {
        progressBar.visibility = View.VISIBLE
        tvMessage.visibility   = View.GONE

        lifecycleScope.launch {
            try {
                val resp = api.listarMinhasOfertas()
                if (resp.isSuccessful) {
                    val dtos = resp.body() ?: emptyList()
                    listaOfertas.clear()
                    listaOfertas.addAll(dtos.map { dto ->
                        Oferta().apply {
                            ofertaId         = dto.id
                            mercadoId        = dto.mercado?.id ?: dto.mercadoId
                            nome             = dto.nome
                            status           = dto.status
                            dataValidade     = dto.data
                            imagemOferta     = dto.imagemOferta ?: dto.imagem
                            nomeSupermercado = dto.mercado?.nome
                        }
                    })
                    ofertaAdapter.setOfertas(listaOfertas)

                    tvMessage.visibility = if (listaOfertas.isEmpty()) {
                        tvMessage.text = "Você ainda não publicou nenhuma oferta."
                        View.VISIBLE
                    } else View.GONE

                } else if (resp.code() == 401) {
                    tratarSessaoExpirada()
                } else {
                    tvMessage.text = "Erro ao carregar (${resp.code()})."
                    tvMessage.visibility = View.VISIBLE
                }
            } catch (e: Exception) {
                Log.e("MinhasOfertas", "Erro ao carregar ofertas", e)
                tvMessage.text = "Sem conexão com o servidor."
                tvMessage.visibility = View.VISIBLE
            } finally {
                progressBar.visibility = View.GONE
            }
        }
    }

    // --- AÇÕES DO ADAPTER ---

    override fun onPdfClick(oferta: Oferta) {
        if (!oferta.imagemOferta.isNullOrEmpty()) {
            startActivity(Intent(this, VerPDF::class.java).apply {
                putExtra("pdfUrl",       oferta.imagemOferta)
                putExtra("oferta_id",    oferta.ofertaId)
                putExtra("oferta_nome",  oferta.nome)
            })
        }
    }

    override fun onEditClick(oferta: Oferta) {
        startActivity(Intent(this, PublicarOfertaActivity::class.java).apply {
            putExtra("oferta_id",     oferta.ofertaId)
            putExtra("oferta_nome",   oferta.nome)
            putExtra("oferta_status", oferta.status)
            putExtra("oferta_data",   oferta.dataValidade)
        })
    }

    override fun onDeleteClick(oferta: Oferta) {
        AlertDialog.Builder(this)
            .setTitle("Excluir Oferta")
            .setMessage("Tem certeza que deseja excluir '${oferta.nome}'?")
            .setPositiveButton("Excluir") { _, _ -> deletarOfertaApi(oferta) }
            .setNegativeButton("Cancelar", null)
            .show()
    }

    override fun onToggleStatusClick(oferta: Oferta) {
        // Implementação futura para Ativar/Desativar rapidamente
        Toast.makeText(this, "Funcionalidade de status em breve", Toast.LENGTH_SHORT).show()
    }

    private fun deletarOfertaApi(oferta: Oferta) {
        val id = oferta.ofertaId ?: return
        lifecycleScope.launch {
            try {
                val resp = api.deletarOferta(id)
                if (resp.isSuccessful) {
                    Toast.makeText(this@MinhasOfertas, "Oferta excluída!", Toast.LENGTH_SHORT).show()
                    carregarOfertas()
                } else {
                    Toast.makeText(this@MinhasOfertas, "Erro ao excluir", Toast.LENGTH_SHORT).show()
                }
            } catch (e: Exception) {
                Toast.makeText(this@MinhasOfertas, "Erro de conexão", Toast.LENGTH_SHORT).show()
            }
        }
    }

    private fun tratarSessaoExpirada() {
        AuthManager.clearSession(this)
        startActivity(Intent(this, LoginActivity::class.java).apply {
            addFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK)
        })
        finish()
    }

    private fun configurarBottomNav() {
        bottomNavigationView.setOnItemSelectedListener { item ->
            when (item.itemId) {
                R.id.navigation_dashboard -> {
                    startActivity(Intent(this, DashboardSupermercadoActivity::class.java))
                    finish(); true
                }
                R.id.navigation_minhas_ofertas -> true
                R.id.navigation_criar_oferta -> {
                    startActivity(Intent(this, PublicarOfertaActivity::class.java))
                    true
                }
                R.id.navigation_gerenciar_ofertas_supermercado -> true
                R.id.navigation_encartes -> {
                    abrirMeusEncartes()
                    true
                }
                R.id.navigation_perfil_supermercado -> {
                    startActivity(Intent(this, Perfil::class.java)); true
                }
                else -> false
            }
        }
        bottomNavigationView.selectedItemId = R.id.navigation_minhas_ofertas
    }

    private fun abrirMeusEncartes() {
        val mercadoId = AuthManager.getUserId(this) ?: return
        startActivity(Intent(this, EncartesActivity::class.java).apply {
            putExtra("mercado_id", mercadoId)
            putExtra("mercado_nome", AuthManager.getNome(this@MinhasOfertas) ?: "Meus encartes")
        })
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) { finish(); return true }
        return super.onOptionsItemSelected(item)
    }
}
